textacy/fileio/utils.py

```diff
@@ -0,0 +1,203 @@
+            or an iterable of lists, e.g. as streamed from a Wikipedia database dump::
+
+                >>> ([pageid, title, text] for pageid, title, text in
+                ...  textacy.corpora.wikipedia.get_plaintext_pages(<WIKIFILE>))
+
+        content_field (str or int): if str, key in each dict item whose value is
+            the item's content (text); if int, index of the value in each list item
+            corresponding to the item's content (text)
+        itemwise (bool, optional): if True, content + metadata are paired item-wise
+            as an iterable of (content, metadata) 2-tuples; if False, content +
+            metadata are paired by position in two parallel iterables in the form of
+            a (iterable(content), iterable(metadata)) 2-tuple
+
+    Returns:
+        generator(tuple(str, dict)): if ``itemwise`` is True and ``items`` is an
+            iterable of dicts; the first element in each tuple is the item's content,
+            the second element is its metadata as a dictionary
+        generator(tuple(str, list)): if ``itemwise`` is True and ``items`` is an
+            iterable of lists; the first element in each tuple is the item's content,
+            the second element is its metadata as a list
+        tuple(iterable(str), iterable(dict)): if ``itemwise`` is False and ``items``
+            is an iterable of dicts; the first element of the tuple is an iterable
+            of items' contents, the second is an iterable of their metadata dicts
+        tuple(iterable(str), iterable(list)): if ``itemwise`` is False and ``items``
+            is an iterable of lists; the first element of the tuple is an iterable
+            of items' contents, the second is an iterable of their metadata lists
+    """
+    if itemwise is True:
+        return ((item.pop(content_field), item) for item in items)
+    else:
+        return unzip(((item.pop(content_field), item) for item in items))
+
+
+def unzip(seq):
+    """
+    Borrowed from ``toolz.sandbox.core.unzip``, but using cytoolz instead of toolz
+    to avoid the additional dependency.
+    """
+    seq = iter(seq)
+    # check how many iterators we need
+    try:
+        first = tuple(next(seq))
+    except StopIteration:
+        return tuple()
+    # and create them
+    niters = len(first)
+    seqs = tee(cons(first, seq), niters)
+    return tuple(starmap(pluck, enumerate(seqs)))
+
+
+def get_filenames(dirname, match_substr=None, ignore_substr=None,
+                  extension=None, ignore_invisible=True, recursive=False):
+    """
+    Yield the full paths of files on disk under directory ``dirname``, optionally
+    filtering for or against particular substrings or file extensions. and crawling all subdirectories.
+
+    Args:
+        dirname (str): /path/to/dir on disk where files to read are saved
+        match_substr (str, optional): match only files with given substring
```

Comment:
Maybe support regex?  Perhaps, not now, but something to put on the roadmap for a future update